tests: error_hook divide by 0 with non-optimized code

FRASTM opened the pull request:
This change makes the test `tests/ztest/error_hook/` pass on stm32 mcu based M0+
nucleo_g071rb

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/35540

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
log of this part:
```
START - test_catch_fatal_error
case type is 0
E: ***** HARD FAULT *****
E: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000000
E: r3/a4:  0x200005e8 r12/ip:  0x0800b6f1 r14/lr:  0x080009e5                   
E:  xpsr:  0x41000000                                                           
E: Faulting instruction address (r15/pc): 0x080007c8                            
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0                             
E: Current thread: 0x20000100 (unknown)                                         
Caught system error -- reason 0 1                                               
Fatal error expected as part of test case.                                      
case type is 1                                                                  
E: ***** HARD FAULT *****                                                       
E: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000000                   
E: r3/a4:  0x20001bd4 r12/ip:  0x0800b6f1 r14/lr:  0x0800b657                   
E:  xpsr:  0x00000000                                                           
E: Faulting instruction address (r15/pc): 0x20001bd4                            
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0                             
E: Current thread: 0x20000100 (unknown)                                         
Caught system error -- reason 0 1                                               
Fatal error expected as part of test case.                                      
case type is 2                                                                  
E: ***** HARD FAULT *****                                                       
E: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000000                   
E: r3/a4:  0x20003000 r12/ip:  0x0800b6f1 r14/lr:  0x08000a01                   
E:  xpsr:  0x41000000                                                           
E: Faulting instruction address (r15/pc): 0x080007f8                            
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0                             
E: Current thread: 0x20000100 (unknown)                                         
Caught system error -- reason 0 1                                               
Fatal error expected as part of test case.                                      
case type is 3                                                                  
E: r0/a1:  0x00000004  r1/a2:  0x00000000  r2/a3:  0x00000000                   
E: r3/a4:  0x00000000 r12/ip:  0x0800b6f1 r14/lr:  0x0800b6d3                   
E:  xpsr:  0x01000000                                                           
E: Faulting instruction address (r15/pc): 0x08000a0e                            
E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0                               
E: Current thread: 0x20000100 (unknown)                                         
Caught system error -- reason 3 1                                               
Fatal error expected as part of test case.                                      
case type is 4                                                                  
E: r0/a1:  0x00000003  r1/a2:  0x00000000  r2/a3:  0x00000000                   
E: r3/a4:  0x00000000 r12/ip:  0x0800b6f1 r14/lr:  0x0800b6d3                   
E:  xpsr:  0x01000000                                                           
E: Faulting instruction address (r15/pc): 0x08000a0e                            
E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0                               
E: Current thread: 0x20000100 (unknown)                                         
Caught system error -- reason 3 1                                               
Fatal error expected as part of test case.                                      
 PASS - test_catch_fatal_error in 0.178 seconds        
```

enjiamai commented:
Hi @FRASTM ,  here is the log after I tested it in k64f_fdrm(cortex-m4):

before applying:
```
DEBUG   - DEVICE: case type is 2
DEBUG   - DEVICE: E: ***** USAGE FAULT *****
DEBUG   - DEVICE: E:   Division by zero
DEBUG   - DEVICE: E: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000001
DEBUG   - DEVICE: E: r3/a4:  0x00000000 r12/ip:  0x0000ab15 r14/lr:  0x00000a3f
DEBUG   - DEVICE: E:  xpsr:  0x61000000
DEBUG   - DEVICE: E: Faulting instruction address (r15/pc): 0x0000085a
DEBUG   - DEVICE: E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
DEBUG   - DEVICE: E: Current thread: 0x200000a0 (unknown)
DEBUG   - DEVICE: Caught system error -- reason 0 1
DEBUG   - DEVICE: Fatal error expected as part of test case.
```

after applying:
```
DEBUG   - DEVICE: case type is 2
DEBUG   - DEVICE: E: ***** BUS FAULT *****
DEBUG   - DEVICE: E:   Precise data bus error
DEBUG   - DEVICE: E:   BFAR Address: 0x20002b80
DEBUG   - DEVICE: E:   NXP MPU error, port 3
DEBUG   - DEVICE: E:     Mode: User, Data Address: 0x20002b80
DEBUG   - DEVICE: E:     Type: Read, Master: 0, Regions: 0x8100
DEBUG   - DEVICE: E: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000000
DEBUG   - DEVICE: E: r3/a4:  0x20002b80 r12/ip:  0x0000ab1d r14/lr:  0x00000a47
DEBUG   - DEVICE: E:  xpsr:  0x21000000
DEBUG   - DEVICE: E: Faulting instruction address (r15/pc): 0x00000850
DEBUG   - DEVICE: E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
DEBUG   - DEVICE: E: Current thread: 0x200000a0 (unknown)
DEBUG   - DEVICE: Caught system error -- reason 0 1
DEBUG   - DEVICE: Fatal error expected as part of test case.
```

It seems like it might change something in other ARM-cortex soc which using "sdiv" instead of  __divsi3 function.
I think this change is good, but still have little concern about affecting the test result of other boards. 
Do you think would it be better to adding a #ifdef to apply this only for nucleo_g071rb or M0+ series?  Thank you!

enjiamai reviewed (commented):
How do you think only apply this change for the testing failed board at this moment? thanks!

FRASTM commented:
yes I agree to restrain this to the M0+ soc series CPU_CORTEX_M0PLUS 
The question is should this be combined with SOC_FAMILY_STM32 ?

FRASTM commented:
Change for M0+ stm32 devices and limiting the impact on the testcase application.

erwango commented on tests/ztest/error_hook/src/main.c:
```diff
@@ -98,10 +98,15 @@ __no_optimization static void trigger_fault_access(void)
 __no_optimization static void trigger_fault_divide_zero(void)
 {
 	int a = 1;
+#if defined(CONFIG_CPU_CORTEX_M0PLUS) && defined(CONFIG_SOC_FAMILY_STM32)
```
Is there a reason to limit this to STM32 ?
@hakehuang how is this test behaving on CM0+ based NXP targets ?

FRASTM replied:
see generated code :
```
   0x08000800 <+16>:	ldr	r3, [pc, #32]	; (0x8000824 <trigger_fault_divide_zero+52>)
   0x08000802 <+18>:	ldr	r3, [r3, #0]
   0x08000804 <+20>:	movs	r1, r3
   0x08000806 <+22>:	ldr	r0, [r7, #4]
   0x08000808 <+24>:	bl	0x800030c <__divsi3>
```

erwango commented:
^^@str4t0m

FRASTM commented:
apply to all cortex M0+ mcus (not stm32 only)
remove static variable init

enjiamai reviewed (approved):
No problem from my side, thanks!

Sorry for bugging you,  @erwango @hakehuang @ABOSTM if you have a chance to verify on other M0+ boards, thank you so much!

ABOSTM reviewed (changes requested):
What about CortexM0 ? I think behavior will be the same than CortexM0+
@FRASTM can you check on STM32F0 ?

FRASTM commented:
No change applied on the cortex M0, as the test did not failed

str4t0m commented:
Thanks @FRASTM.
I can also only test on a STM32G0.

The test now passes, but I don't think the fault happens in the division, but already before.
For me it happens at instruction 0x080007fe.

`080007FA   LDR            R3, =b.1                      ; [PC, #40] [0x08000824] =0x200003E8

 080007FC   MOVS           R2, #0

 080007FE   STR            R2, [R3, #0]`

The test also passes, due to the fault, if the division is replaced by an addition.

str4t0m reviewed (changes requested):
Further analysing #35540 I believe that that __aeabi_idiv is correctly called.
__aeabi_idiv includes a check if the quotient is zero, as this is the case __aeabi_div0 is called.
Per default __aeabi_div0 simply returns, such that result 0 is returned from division.

So I think there are two possibilities to resolve this:
- Accept that this targets don't cause a fault and instead return 0:
  Add CONFIG_CPU_CORTEX_M0PLUS (and probably CONFIG_CPU_CORTEX_M0) to targets that skip the division test.
- Overwrite the __aeabi_div0 from the standard library to trigger a fault.

I believe non STM32 targets should have the same issue, so it would be interesting to have feedback from such targets.

str4t0m commented:
As an alternative the test passes without this modifications if I add the following sketch in arch/arm/core/aarch32/cortex_m/fault.c:

```
#if defined(CONFIG_CPU_CORTEX_M0PLUS)
void __aeabi_idiv0(void)
{
	printk("Div zero\n");
	k_panic();
}
#endif /* CONFIG_CPU_CORTEX_M0PLUS */
```

But I am not sure whether it is preferable to cause a fault or return 0.

enjiamai commented:
> So I think there are two possibilities to resolve this:
> 
> * Accept that this targets don't cause a fault and instead return 0:
>   Add CONFIG_CPU_CORTEX_M0PLUS (and probably CONFIG_CPU_CORTEX_M0) to targets that skip the division test.
> * Overwrite the __aeabi_div0 from the standard library to trigger a fault.

I think @str4t0m is right, the exception trigger is due to the access to the variable "volatile int b", although the testcase passes.  

Excuse me, @FRASTM, could you please apply the solution of what @str4t0m has recommended? Both of them look better from my point of view. 

Sorry, @hakehuang , thanks so much for your help with verifying this.

FRASTM commented:
Test passed on nucleo_g071rb (cortexM0+) without changing the testcase itself when adding this code to the  arch/arm/core/aarch32/cortex_m/fault.c:
```
#if defined(CONFIG_CPU_CORTEX_M0PLUS)
void __aeabi_idiv0(void)
{
	printk("Div zero\n");
	k_panic();
}
#endif /* CONFIG_CPU_CORTEX_M0PLUS */
```

enjiamai commented:
Hi, @FRASTM, sounds great. I suggest just skip them and make some descriptions for them.

FRASTM commented:
@enjiamai, do you mean changing the `arch/arm/core/aarch32/cortex_m/fault.c` instead of `tests/ztest/error_hook/src/main.c `for the any CONFIG_CPU_CORTEX_M0PLUS ?

enjiamai commented:
> @enjiamai, do you mean changing the `arch/arm/core/aarch32/cortex_m/fault.c` instead of `tests/ztest/error_hook/src/main.c `for the any CONFIG_CPU_CORTEX_M0PLUS ?

I prefer to modify the testcase to skip it, like this:
```
diff --git a/tests/ztest/error_hook/src/main.c b/tests/ztest/error_hook/src/main.c
index f6d8e91fbf..0c0c74c284 100644
--- a/tests/ztest/error_hook/src/main.c
+++ b/tests/ztest/error_hook/src/main.c
@@ -112,7 +112,8 @@ __no_optimization static void trigger_fault_divide_zero(void)
  * test will still apply on the physical board.
  */
 #if (defined(CONFIG_SOC_SERIES_MPS2) && (CONFIG_QEMU_TARGET)) || \
-       (CONFIG_BOARD_QEMU_CORTEX_A53) || (CONFIG_SOC_QEMU_ARC)
+       (CONFIG_BOARD_QEMU_CORTEX_A53) || (CONFIG_SOC_QEMU_ARC) || \
+       (CONFIG_CPU_CORTEX_M0PLUS)
        ztest_test_skip();
 #endif
 }
```
I think it would be a lower risk. IMO.
Then we can check if M0+ has something like "[Configurable Fault Status Register](https://developer.arm.com/documentation/dui0552/a/cortex-m3-peripherals/system-control-block/configurable-fault-status-register?lang=en)" to configure and make a change/fix after the v2.6 release. (Sorry I don't really understand ARM very much)

hakehuang commented:
> Sorry for bugging you, @erwango @hakehuang @ABOSTM if you have a chance to verify on other M0+ boards, thank you so much!

in NXP frdm_kw41z

```
*** Booting Zephyr OS build v2.6.0-rc1-202-g6f8ddc4c50b5  ***
Running test suite error_hook_tests
===================================================================
START - test_catch_fatal_error
case type is 3
E: r0/a1:  0x00000004  r1/a2:  0x0000000f  r2/a3:  0x00000000
E: r3/a4:  0x200003d8 r12/ip:  0x00000003 r14/lr:  0x00000b15
E:  xpsr:  0x21000000
E: Faulting instruction address (r15/pc): 0x00004a4c
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x20000140 (unknown)
Caught system error -- reason 4 1
Fatal error expected as part of test case.
case type is 4
E: r0/a1:  0x00000003  r1/a2:  0x0000000f  r2/a3:  0x00000000
E: r3/a4:  0x200003d8 r12/ip:  0x00000004 r14/lr:  0x00000b21
E:  xpsr:  0x21000000
E: Faulting instruction address (r15/pc): 0x00004a44
E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
E: Current thread: 0x20000140 (unknown)
Caught system error -- reason 3 1
Fatal error expected as part of test case.
 PASS - test_catch_fatal_error in 0.69 seconds
===================================================================
START - test_catch_assert_fail
ASSERTION FAIL [a != ((void *)0)] @ ../src/main.c:41
        parameter a should not be NULL!
Caught assert failed
Assert error expected as part of test case.
 PASS - test_catch_assert_fail in 0.14 seconds
===================================================================
START - test_catch_assert_in_isr
ASSERTION FAIL [a != ((void *)0)] @ ../src/main.c:41
        parameter a should not be NULL!
Caught assert failed
Assert error expected as part of test case.
 PASS - test_catch_assert_in_isr in 0.14 seconds
===================================================================
Test suite error_hook_tests succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL

```

enjiamai commented:
Hi, @FRASTM,  any update or idea for this? thanks!

FRASTM commented:
Sorry, I do not catch what is expected as a conclusion :  Skip the test for M0+ is also possible to "PASS", of course or something else

ioannisg reviewed (changes requested):
You should probably provide a brief explanation in the commit message on the solution that is picked, and a more detailed inline explanation in the test code. Right now, I cannot easily see the root cause of this

ioannisg commented:
And, by the way, skipping the test is not good; Cortex-m0+ should be supported fully. So I'd be in favor of the current solution, if it works.

enjiamai commented:
Hi, @ioannisg ,  the original testcase failed is due to an excepted exception did not be triggered on M0+ based nucleo_g071rb.
I guess on this board the "trigger fault of divided by zero" might not be configured, but not very sure.
 
The current solution will trigger an exception but not due to "divided by zero".  That's why I suggest just skip and make some descriptions for it.

ioannisg commented:
OK I see, now what you mean @enjiamai. I had a closer look now. Of course, Cortex-M baseline instruction set does not seem to have the divide instructions, so they do not have the corresponding exception either. 

See: https://stackoverflow.com/questions/41856638/divide-by-zero-exception-on-m0-devices

It is up to the compiler to trigger an undefined instruction. But we should not be testing the compiler here. We should be testing Zephyr :) So I think we can just skip the test case for Cortex-M baselines. This includes Cortex-M0 and Cortex-M23. So better to use ARMV6_M_ARMV8_M_BASELINE Kconfig for this. 

You can add a test case, where ARM cores explicitly insert an Undefined Instruction, and you verify the fault. But not in the test case for divide-by-zero, IMHO.

ioannisg commented:
> Further analysing #35540 I believe that that __aeabi_idiv is correctly called.

@str4t0m that's all compiler stuff, which we should not test, IMHO.
Strictly speaking we should not even test the hardware here :) But at least for the Cortex-M Mainline, we _do_ set in the core initialization the DIV0 trap, so having the test here, has some value :)

ioannisg commented:
@galak that's one more low-hanging fruit for the 2.6 release.

enjiamai commented:
>Strictly speaking we should not even test the hardware here :) But at least for the Cortex-M Mainline, we do set in the core >initialization the DIV0 trap, so having the test here, has some value :)

Excuse me, @ioannisg , one more question :)
Do you means beside skip it, we can also apply the solution of modifying the arch/arm/core/aarch32/cortex_m/fault.c like this?
```
#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE) && defined(CONFIG_TEST)
void __aeabi_idiv0(void)
{
	printk("Div zero\n");
	k_panic();
}
#endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE  && CONFIG_TEST */
```
Thank you!

FRASTM committed:
tests: error_hook skipping div-by-0 test on cortex M0 M0+

On the Cortex M0+, M0, M23 the div-by-zero is not mapped
as an hardware exception but triggers a HardFault.
The test case is skipped for any ARMV6_M_ARMV8_M_BASELINE.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
skip this case with ARMV6_M_ARMV8_M_BASELINE

enjiamai reviewed (approved):
It would be better if Cortex-M23 can be added to the comments, thank you.

ioannisg reviewed (approved):
Thanks @FRASTM

ioannisg commented:
> It would be better if Cortex-M23 can be added to the comments, thank you.

CM32 is part of the Baseline Cortex-M, anyways.

FRASTM commented:
changing comment to include cortex-M23

ABOSTM reviewed (commented):
Commit message is not clear: 

> On the Cortex M0+, M0, M23 the div-by-zero is not mapped
> as an exception but triggers a HardFault.

Hardfault is an exception.

ABOSTM commented:
> > Commit message is not clear:
> > > On the Cortex M0+, M0, M23 the div-by-zero is not mapped
> > > as an exception but triggers a HardFault.
> > 
> > 
> > Hardfault is an exception.
> 
> It's not a dedicated exception, e.g. due to a divide-by-zero trap. Maybe add this detail in the commit body, if we're to be a bit more pedantic

I still think it is ambiguous as hardfault is an exception (as defined by ARM), even if it is not a dedicated one.

ioannisg commented:
> I am not sure to understand why we want to skip the test for ARMV6_M_ARMV8_M_BASELINE:
> As I understand with such Arch, there is no dedicated exception, because there is no division instruction.
> But I also understand that in this case Zephyr generate an Hardfault.

This depends on the compiler, not Zephyr. So I argued that we do not need to be testing this.
In theory we could have a compiler that does not generate an undefined instruction that leads to a HF.

ABOSTM commented:
Ok thanks for the clarification I understand you point, but it means behavior is compiler dependent !!! 
That sounds weird and dangerous from application point of view.
And if no fault is generated, the consequences of retuning 0 for the div instead of fault are unpredictable and potentially harmful :-(

ioannisg commented:
> Ok thanks for the clarification I understand you point, but it means behavior is compiler dependent !!!
> That sounds weird and dangerous from application point of view.
> And if no fault is generated, the consequences of retuning 0 for the div instead of fault are unpredictable and potentially harmful :-(

I do not know the details of GCC spec about this, so it was only a wild assumption that there could be a compiler without generating Undefined Instruction. My point was that Zephyr testing suite should not be verifying the compiler behavior; we should verify the software we write.

There is nothing about the software we write, here, for Cortex-M Baseline. We rely on the compiler generating an UDF instruction.

For Mainline, it is a different story. there is an ARM configuration register bit that controls whether we catch DIV0 traps as UsgFaults or not. Therefore, the value of this test here, is to guard that we do not alter that selected configuration (so it is about "software we write").

ABOSTM commented:
Ok, I am not fully convinced, as even is it is compiler and not the "software we write" there is still undetermined behavior.
But I will follow the majority.

ABOSTM reviewed (approved):
LGTM

enjiamai commented:
This divide_by_zero testcase at first actually generate UDF in userspace, then it was modified by adding a __no_optimization (in order to address #33403). 
Maybe we can add testcases that show the behavior of both without __no_optimization and non-userspace in the future.
